Sources/BigInt/BigInt.swift

```diff
@@ -330,37 +330,29 @@ extension BigInt: SignedNumeric {
 extension BigInt: BinaryInteger {
 
   public init?<T>(exactly source: T) where T: BinaryFloatingPoint {
-    if (source.isNaN || source.isInfinite) ||
-      (source.rounded(.towardZero) != source) {
+    guard source.isFinite, source == source.rounded(.towardZero) else {
       return nil
     }
-
     self.init(source)
   }
 
   public init<T>(_ source: T) where T: BinaryFloatingPoint {
-    precondition(
-      !(source.isNaN || source.isInfinite),
-      "\(type(of: source)) value cannot be converted to BigInt because it is either infinite or NaN"
-    )
-
-    let isNegative = source < 0.0
-    var float = isNegative ? -source : source
+    precondition(source.isFinite, "BigInt(\(source)) failed")
+    precondition(T.significandBitCount < .max)
 
-    if let _ = UInt(exactly: T.greatestFiniteMagnitude) {
-      words = [UInt(float)]
-    } else {
-      var words = Words()
-      let radix = T(sign: .plus, exponent: T.Exponent(UInt.bitWidth), significand: 1)
-      repeat {
-        let digit = UInt(float.truncatingRemainder(dividingBy: radix))
-        words.append(digit)
-        float = (float / radix).rounded(.towardZero)
-      } while float != 0
+    let source = source.rounded(.towardZero)
+    let isNegative = source < 0
 
-      self.words = words
+    guard !source.isZero else {
+      self = 0
+      return
     }
 
+    self = BigInt(source.significandBitPattern)
+    self |= BigInt(1) << T.significandBitCount
```

Comment:
So I don't think `BigInt` shifts are actually going to be faster than floating-point division. Though that could change if someone makes a really big `BinaryFloatingPoint` type, I suppose. I guess this is more immediately clear as to what's going on.

Reply:
I only changed this part because a positive `greatestFiniteMagnitude` was converting into a negative BigInt.

The bit-shifting is clearer to me than the division-based code, but I didn't realize that it might be slower.

Reply:
Right, because the bit shifting is actually calling the BigInt functions to do that, instead of using the floating-point hardware haha. But obviously that goes out the window in the case of a `BinaryFloatingPoint` that isn't a `Float32`, `Float64`, or `Float80`. So it mostly depends on which case we're optimizing for… and if we even need to care. I think hopefully people won't convert floating-point types to `BigInt` in some tight inner loop, so probably the bit-shifting is clearer and better even if it's slower in many cases. It's quite possibly better for memory usage, too, depending. That said, for handling the case of a flexible `BinaryFloatingPoint` number with no upper limit to the size of the significand, I think we have to use the division code, as I mention elsewhere.